pkg/backup/backupschedule/backup_schedule_manager.go

```diff
@@ -288,26 +357,232 @@ func (bm *backupScheduleManager) backupGCByMaxReservedTime(bs *v1alpha1.BackupSc
 		return
 	}
 
-	var deleteCount int
-	for _, backup := range backupsList {
-		if backup.CreationTimestamp.Add(reservedTime).After(bm.now()) {
-			continue
+	ascBackups, logBackup := separateSnapshotBackupsAndLogBackup(backupsList)
+	if len(ascBackups) == 0 {
+		return
+	}
+
+	var (
+		expiredBackups []*v1alpha1.Backup
+		truncateTSO    uint64
+	)
+
+	if logBackup != nil {
+		expiredBackups, truncateTSO, err = caculateExpiredBackupsWithLogBackup(ascBackups, logBackup, reservedTime)
+		if err != nil {
+			klog.Errorf("caculate expired backups with log backup, err: %s", err)
+			return
 		}
+	} else {
+		expiredBackups, err = caculateExpiredBackups(ascBackups, reservedTime)
+		if err != nil {
+			klog.Errorf("caculate expired backups without log backup, err: %s", err)
+			return
+		}
+	}
+
+	for _, backup := range expiredBackups {
 		// delete the expired backup
-		if err := bm.deps.BackupControl.DeleteBackup(backup); err != nil {
+		if err = bm.deps.BackupControl.DeleteBackup(backup); err != nil {
 			klog.Errorf("backup schedule %s/%s gc backup %s failed, err %v", ns, bsName, backup.GetName(), err)
 			return
 		}
-		deleteCount += 1
 		klog.Infof("backup schedule %s/%s gc backup %s success", ns, bsName, backup.GetName())
 	}
 
-	if deleteCount == len(backupsList) && deleteCount > 0 {
+	if truncateTSO > 0 {
+		// truncate the log backup
+		if err = bm.deps.BackupControl.TruncateLogBackup(logBackup, truncateTSO); err != nil {
+			klog.Errorf("backup schedule %s/%s truncate log backup %s failed, truncateTSO %d, err %v", ns, bsName, logBackup.GetName(), truncateTSO, err)
+			return
+		}
+		klog.Infof("backup schedule %s/%s truncate log backup %s success, truncateTSO %d", ns, bsName, logBackup.GetName(), truncateTSO)
+	}
+
+	if len(expiredBackups) == len(backupsList) && len(expiredBackups) > 0 {
 		// All backups have been deleted, so the last backup information in the backupSchedule should be reset
 		bm.resetLastBackup(bs)
 	}
 }
 
```

Comment:
The following functions only have pure math calculation logic, not any service calls, please write unit test for them

Reply:
ok